File: KNIRVENGINE/desktop-client/services/knirvoracleService.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// KNIRVOracleService provides integration with KNIRVORACLE for agent minting,
// capability registration, wallet operations, and network interactions
type KNIRVOracleService struct {
	baseURL    string
	httpClient *http.Client
	apiKey     string
}

// Configuration for KNIRVORACLE service
type KNIRVOracleConfig struct {
	BaseURL string `json:"base_url"`
	APIKey  string `json:"api_key"`
	Timeout int    `json:"timeout_seconds"`
}

// Agent minting request structure
type AgentMintRequest struct {
	AgentID     string                 `json:"agent_id"`
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Owner       string                 `json:"owner"`
	Metadata    map[string]interface{} `json:"metadata"`
	ImageURL    string                 `json:"image_url,omitempty"`
}

// Agent minting response structure
type AgentMintResponse struct {
	Success       bool   `json:"success"`
	TransactionID string `json:"transaction_id"`
	AgentNFTID    string `json:"agent_nft_id"`
	Message       string `json:"message"`
}

// Capability registration request structure
type CapabilityRegistrationRequest struct {
	Name          string                 `json:"name"`
	Type          string                 `json:"type"`
	Description   string                 `json:"description"`
	Schema        map[string]interface{} `json:"schema"`
	Owner         string                 `json:"owner"`
	GasFeeNRN     uint64                 `json:"gas_fee_nrn"`
	LocationHints []string               `json:"location_hints,omitempty"`
}

// Capability registration response structure
type CapabilityRegistrationResponse struct {
	Success      bool   `json:"success"`
	CapabilityID string `json:"capability_id"`
	TxHash       string `json:"tx_hash"`
	Message      string `json:"message"`
}

// Faucet request structure
type FaucetRequest struct {
	Address string `json:"address"`
	Amount  string `json:"amount"`
	Reason  string `json:"reason,omitempty"`
}

// Faucet response structure
type FaucetResponse struct {
	Success   bool   `json:"success"`
	RequestID string `json:"request_id"`
	TxHash    string `json:"tx_hash"`
	Amount    string `json:"amount"`
	Status    string `json:"status"`
	Message   string `json:"message"`
}

// Wallet balance response structure
type WalletBalanceResponse struct {
	Success    bool   `json:"success"`
	Address    string `json:"address"`
	Balance    string `json:"balance"`
	NRNBalance string `json:"nrn_balance"`
	USDValue   string `json:"usd_value"`
}

// Transaction request structure
type TransactionRequest struct {
	From   string `json:"from"`
	To     string `json:"to"`
	Amount string `json:"amount"`
	Token  string `json:"token"`
	Memo   string `json:"memo,omitempty"`
	GasFee string `json:"gas_fee,omitempty"`
}

// Transaction response structure
type TransactionResponse struct {
	Success       bool   `json:"success"`
	TransactionID string `json:"transaction_id"`
	TxHash        string `json:"tx_hash"`
	Status        string `json:"status"`
	Message       string `json:"message"`
}

// NewKNIRVOracleService creates a new KNIRVORACLE service instance
func NewKNIRVOracleService(config KNIRVOracleConfig) *KNIRVOracleService {
	timeout := time.Duration(config.Timeout) * time.Second
	if timeout == 0 {
		timeout = 30 * time.Second
	}

	return &KNIRVOracleService{
		baseURL: config.BaseURL,
		apiKey:  config.APIKey,
		httpClient: &http.Client{
			Timeout: timeout,
		},
	}
}

// makeRequest performs HTTP requests to KNIRVORACLE with proper error handling
func (k *KNIRVOracleService) makeRequest(ctx context.Context, method, endpoint string, body interface{}) (*http.Response, error) {
	var reqBody io.Reader
	if body != nil {
		jsonData, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("failed to marshal request body: %w", err)
		}
		reqBody = bytes.NewBuffer(jsonData)
	}

	url := k.baseURL + endpoint
	req, err := http.NewRequestWithContext(ctx, method, url, reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "KNIRVENGINE-Desktop-Client/1.0")
	if k.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+k.apiKey)
	}

	resp, err := k.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}

	return resp, nil
}

// MintAgent mints a new agent as an NFT on KNIRVORACLE
func (k *KNIRVOracleService) MintAgent(ctx context.Context, req AgentMintRequest) (*AgentMintResponse, error) {
	resp, err := k.makeRequest(ctx, "POST", "/agent/mint", req)
	if err != nil {
		return nil, fmt.Errorf("failed to mint agent: %w", err)
	}
	defer resp.Body.Close()

	var result AgentMintResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return &result, fmt.Errorf("agent minting failed: %s", result.Message)
	}

	return &result, nil
}

// RegisterCapability registers a capability with KNIRVORACLE
func (k *KNIRVOracleService) RegisterCapability(ctx context.Context, req CapabilityRegistrationRequest) (*CapabilityRegistrationResponse, error) {
	resp, err := k.makeRequest(ctx, "POST", "/wallet/mcp/create_register_capability", req)
	if err != nil {
		return nil, fmt.Errorf("failed to register capability: %w", err)
	}
	defer resp.Body.Close()

	var result CapabilityRegistrationResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return &result, fmt.Errorf("capability registration failed: %s", result.Message)
	}

	return &result, nil
}

// RequestFaucet requests tokens from the KNIRVORACLE faucet
func (k *KNIRVOracleService) RequestFaucet(ctx context.Context, req FaucetRequest) (*FaucetResponse, error) {
	resp, err := k.makeRequest(ctx, "POST", "/api/mint/nrv", req)
	if err != nil {
		return nil, fmt.Errorf("failed to request faucet: %w", err)
	}
	defer resp.Body.Close()

	var result FaucetResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return &result, fmt.Errorf("faucet request failed: %s", result.Message)
	}

	return &result, nil
}

// GetWalletBalance retrieves wallet balance from KNIRVORACLE
func (k *KNIRVOracleService) GetWalletBalance(ctx context.Context, address string) (*WalletBalanceResponse, error) {
	endpoint := fmt.Sprintf("/balance/%s", address)
	resp, err := k.makeRequest(ctx, "GET", endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get wallet balance: %w", err)
	}
	defer resp.Body.Close()

	var result WalletBalanceResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return &result, fmt.Errorf("failed to get balance: status %d", resp.StatusCode)
	}

	return &result, nil
}

// SendTransaction sends a transaction through KNIRVORACLE
func (k *KNIRVOracleService) SendTransaction(ctx context.Context, req TransactionRequest) (*TransactionResponse, error) {
	resp, err := k.makeRequest(ctx, "POST", "/transactions", req)
	if err != nil {
		return nil, fmt.Errorf("failed to send transaction: %w", err)
	}
	defer resp.Body.Close()

	var result TransactionResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return &result, fmt.Errorf("transaction failed: %s", result.Message)
	}

	return &result, nil
}

// GetAgentCapabilities retrieves capabilities for a specific agent
func (k *KNIRVOracleService) GetAgentCapabilities(ctx context.Context, agentID string) ([]map[string]interface{}, error) {
	endpoint := fmt.Sprintf("/agent/capabilities/%s", agentID)
	resp, err := k.makeRequest(ctx, "GET", endpoint, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get agent capabilities: %w", err)
	}
	defer resp.Body.Close()

	var result struct {
		Success      bool                     `json:"success"`
		Capabilities []map[string]interface{} `json:"capabilities"`
		Message      string                   `json:"message"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to get capabilities: %s", result.Message)
	}

	return result.Capabilities, nil
}

// InvokeCapability invokes a capability through KNIRVORACLE
func (k *KNIRVOracleService) InvokeCapability(ctx context.Context, capabilityID string, inputData map[string]interface{}) (map[string]interface{}, error) {
	reqBody := map[string]interface{}{
		"capability_id":    capabilityID,
		"interaction_type": "invoke",
		"input_data":       inputData,
		"timestamp":        time.Now().Unix(),
	}

	resp, err := k.makeRequest(ctx, "POST", "/wallet/mcp/create_invoke_capability", reqBody)
	if err != nil {
		return nil, fmt.Errorf("failed to invoke capability: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return result, fmt.Errorf("capability invocation failed")
	}

	return result, nil
}

// GetTreasuryStatus retrieves treasury status from KNIRVORACLE
func (k *KNIRVOracleService) GetTreasuryStatus(ctx context.Context) (map[string]interface{}, error) {
	resp, err := k.makeRequest(ctx, "GET", "/api/treasury/status", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to get treasury status: %w", err)
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return result, nil
}

// HealthCheck performs a health check against KNIRVORACLE
func (k *KNIRVOracleService) HealthCheck(ctx context.Context) error {
	resp, err := k.makeRequest(ctx, "GET", "/health", nil)
	if err != nil {
		return fmt.Errorf("health check failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("KNIRVORACLE unhealthy: status %d", resp.StatusCode)
	}

	return nil
}

File: packages/KNIRVENGINE/gui/src/services/knirvoracleService.ts
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import {
  KNIRVOracleConfig,
  AgentMintRequest,
  AgentMintResponse,
  CapabilityRegistrationRequest,
  CapabilityRegistrationResponse,
  FaucetRequest,
  FaucetResponse,
  WalletBalanceResponse,
  TransactionRequest,
  TransactionResponse,
  AgentCapability,
  KNIRVOracleError,
  KNIRVORACLE_ENDPOINTS,
  KNIRVORACLE_TIMEOUTS,
  isKNIRVOracleError,
} from '../types/knirvoracle';

// Re-export types for convenience
export type {
  KNIRVOracleConfig,
  AgentMintRequest,
  AgentMintResponse,
  CapabilityRegistrationRequest,
  CapabilityRegistrationResponse,
  FaucetRequest,
  FaucetResponse,
  WalletBalanceResponse,
  TransactionRequest,
  TransactionResponse,
  AgentCapability,
} from '../types/knirvoracle';

/**
 * KNIRVOracleService provides TypeSafe integration with KNIRVORACLE
 * for agent minting, capability registration, wallet operations, and network interactions
 */
export class KNIRVOracleService {
  private client: AxiosInstance;
  private config: KNIRVOracleConfig;

  constructor(config: KNIRVOracleConfig) {
    this.config = config;
    this.client = axios.create({
      baseURL: config.baseURL,
      timeout: config.timeout || KNIRVORACLE_TIMEOUTS.TRANSACTION,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'KNIRVENGINE-Desktop-Client/1.0',
        ...(config.apiKey && { Authorization: `Bearer ${config.apiKey}` }),
      },
    });

    // Add response interceptor for error handling
    this.client.interceptors.response.use(
      (response) => response,
      (error) => {
        const knirvoracleError: KNIRVOracleError = new Error(
          `KNIRVORACLE API Error: ${error.response?.data?.message || error.message}`
        ) as KNIRVOracleError;

        knirvoracleError.isKNIRVOracleError = true;
        knirvoracleError.code = error.response?.data?.error_code;
        knirvoracleError.status = error.response?.status;
        knirvoracleError.response = error.response?.data;

        console.error('KNIRVORACLE API Error:', knirvoracleError);
        return Promise.reject(knirvoracleError);
      }
    );
  }

  /**
   * Mint a new agent as an NFT on KNIRVORACLE
   */
  async mintAgent(request: AgentMintRequest): Promise<AgentMintResponse> {
    try {
      const response: AxiosResponse<AgentMintResponse> = await this.client.post(
        KNIRVORACLE_ENDPOINTS.AGENT_MINT,
        request,
        { timeout: KNIRVORACLE_TIMEOUTS.AGENT_MINT }
      );
      return response.data;
    } catch (error) {
      if (isKNIRVOracleError(error)) {
        throw error;
      }
      throw new Error(`Failed to mint agent: ${error}`);
    }
  }

  /**
   * Register a capability with KNIRVORACLE
   */
  async registerCapability(request: CapabilityRegistrationRequest): Promise<CapabilityRegistrationResponse> {
    try {
      const response: AxiosResponse<CapabilityRegistrationResponse> = await this.client.post(
        KNIRVORACLE_ENDPOINTS.CAPABILITY_REGISTER,
        request,
        { timeout: KNIRVORACLE_TIMEOUTS.CAPABILITY_REGISTER }
      );
      return response.data;
    } catch (error) {
      if (isKNIRVOracleError(error)) {
        throw error;
      }
      throw new Error(`Failed to register capability: ${error}`);
    }
  }

  /**
   * Request tokens from the KNIRVORACLE faucet
   */
  async requestFaucet(request: FaucetRequest): Promise<FaucetResponse> {
    try {
      const response: AxiosResponse<FaucetResponse> = await this.client.post(
        '/api/mint/nrv',
        request
      );
      return response.data;
    } catch (error) {
      throw new Error(`Failed to request faucet: ${error}`);
    }
  }

  /**
   * Get wallet balance from KNIRVORACLE
   */
  async getWalletBalance(address: string): Promise<WalletBalanceResponse> {
    try {
      const response: AxiosResponse<WalletBalanceResponse> = await this.client.get(
        `/balance/${address}`
      );
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get wallet balance: ${error}`);
    }
  }

  /**
   * Send a transaction through KNIRVORACLE
   */
  async sendTransaction(request: TransactionRequest): Promise<TransactionResponse> {
    try {
      const response: AxiosResponse<TransactionResponse> = await this.client.post(
        '/transactions',
        request
      );
      return response.data;
    } catch (error) {
      throw new Error(`Failed to send transaction: ${error}`);
    }
  }

  /**
   * Get capabilities for a specific agent
   */
  async getAgentCapabilities(agentId: string): Promise<AgentCapability[]> {
    try {
      const response: AxiosResponse<{ success: boolean; capabilities: AgentCapability[]; message: string }> = 
        await this.client.get(`/agent/capabilities/${agentId}`);
      
      if (!response.data.success) {
        throw new Error(response.data.message);
      }
      
      return response.data.capabilities;
    } catch (error) {
      throw new Error(`Failed to get agent capabilities: ${error}`);
    }
  }

  /**
   * Invoke a capability through KNIRVORACLE
   */
  async invokeCapability(capabilityId: string, inputData: Record<string, unknown>): Promise<Record<string, unknown>> {
    try {
      const requestBody = {
        capability_id: capabilityId,
        interaction_type: 'invoke',
        input_data: inputData,
        timestamp: Date.now(),
      };

      const response: AxiosResponse<Record<string, unknown>> = await this.client.post(
        '/wallet/mcp/create_invoke_capability',
        requestBody
      );
      
      return response.data;
    } catch (error) {
      throw new Error(`Failed to invoke capability: ${error}`);
    }
  }

  /**
   * Get treasury status from KNIRVORACLE
   */
  async getTreasuryStatus(): Promise<Record<string, unknown>> {
    try {
      const response: AxiosResponse<Record<string, unknown>> = await this.client.get('/api/treasury/status');
      return response.data;
    } catch (error) {
      throw new Error(`Failed to get treasury status: ${error}`);
    }
  }

  /**
   * Perform a health check against KNIRVORACLE
   */
  async healthCheck(): Promise<boolean> {
    try {
      const response = await this.client.get(
        KNIRVORACLE_ENDPOINTS.HEALTH,
        { timeout: KNIRVORACLE_TIMEOUTS.HEALTH_CHECK }
      );
      return response.status === 200;
    } catch (error) {
      console.error('KNIRVORACLE health check failed:', error);
      return false;
    }
  }

  /**
   * Update the service configuration
   */
  updateConfig(newConfig: Partial<KNIRVOracleConfig>): void {
    this.config = { ...this.config, ...newConfig };
    
    // Update axios instance with new config
    if (newConfig.baseURL) {
      this.client.defaults.baseURL = newConfig.baseURL;
    }
    
    if (newConfig.timeout) {
      this.client.defaults.timeout = newConfig.timeout;
    }
    
    if (newConfig.apiKey) {
      this.client.defaults.headers.Authorization = `Bearer ${newConfig.apiKey}`;
    }
  }

  /**
   * Get current configuration
   */
  getConfig(): KNIRVOracleConfig {
    return { ...this.config };
  }
}

// Create a default instance for use throughout the application
export const knirvoracleService = new KNIRVOracleService({
  baseURL: import.meta.env.VITE_KNIRVORACLE_URL || 'http://localhost:8080',
  apiKey: import.meta.env.VITE_KNIRVORACLE_API_KEY,
  timeout: 30000,
});

export default knirvoracleService;
